File: Week 1- Programming Challenges/1. Maximum pairwise product/max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <cstdlib>

using std::vector;
using std::cin;
using std::cout;

double MaxPairwiseProductNaive(const vector<int>& numbers) {
  double result = 0;
  int n = numbers.size();
  for (int i = 0; i < n; ++i) {
    for (int j = i + 1; j < n; ++j) {
      if (numbers[i] * numbers[j] > result) {
        result = numbers[i] * numbers[j];
      }
    }
  }
  return result;
}

long long MaxPairwiseProductFast(const vector<int>& numbers){
  int a=0,b=0,j=0;
  int n = numbers.size();
  for (int i = 0; i<n; ++i){
    if(numbers[i]>=a){
      a=numbers[i];
      j=i;
    }
    }
    for (int i = 0; i<n; ++i){
      if(numbers[i]>=b and i!=j){
        b=numbers[i];
  }
}
  // cout<< a << " " << b << "\n";
  return ((long long) a)*b;
}

int main() {
    int n;
    cin >> n;
    vector<int> numbers(n);

    for (int i = 0; i < n; ++i) {
        cin >> numbers[i];
    }

    /// Test case 1:
    // vector<int> testnumbers(200000);
    // for(int i = 0;i<200000;++i){
    //   testnumbers[i]=i+1;
    // }

    /// Teset case 2:
    // while (true) {
    //   int n = rand() % 1000 + 2;
    //   cout << n << "\n";
    //   vector<int> a;
    //   for (int i = 0; i < n; ++i) {
    //     a.push_back(rand() % 100000);
    //   }
    //   for (int i = 0; i < n; ++i) {
    //     cout << a[i] << ' ';
    //   }
    //   cout << "\n";
    //   long long res1 = MaxPairwiseProductNaive(a);
    //   long long res2 = MaxPairwiseProductFast(a);
    //   if (res1 != res2) {
    //     cout << "Wrong answer: " << res1 << ' ' << res2 << "\n";
    //     break;
    //   }
    //   else {
    //     cout << "OK\n";
    //   }
    // }


    long long result=MaxPairwiseProductFast(numbers);
    cout << result << "\n";
    return 0;
}

File: Week 1- Programming Challenges/1. Maximum pairwise product/max_pairwise_product.py
# Uses python3
#import numpy as np

def MaxPairwiseProductNaive(a):
    n= len(a)
    result = 0
#    index1,index2=0,0
    for i in range(0, n):
        for j in range(i+1, n):
            if a[i]*a[j] > result:
                result = a[i]*a[j]
#                index1,index2=i,j
#    print("Naive: num1:",a[index1],"num2:",a[index2])
    return result

def MaxPairwiseProductFast(a):
    num1,num2,index=0,0,0
    n=len(a)
    for i in range(n):
        if(a[i]>=num1):
            index=i
            num1=a[i]
    for i in range(n):
        if(a[i]>=num2 and i!=index):
            num2=a[i]
#    print ("Fast: num1:",num1,"num2:",num2)
    return num1*num2

# Stress Test
#while(True):
#    ran_n=np.random.randint(2,1000,1,dtype='int64')
#    print(ran_n)
#    ran_arr=np.random.randint(1,100000,ran_n,dtype='int64')
#    print(ran_arr)
#    res1=MaxPairwiseProductNaive(ran_arr)
#    res2=MaxPairwiseProductFast(ran_arr)
#
#    if (res1!=res2):
#        print("Wrong Answer")
#        print(res1,res2)
#        break
#    else: print("OK")

# Main Program
n = int(input())
a = [int(x) for x in input().split()]
assert(len(a) == n)

print(MaxPairwiseProductFast(a))
